File: convex/mafengwo.ts
/**
 * 马蜂窝数据存储 Mutations
 * 处理各类马蜂窝爬取数据的存储
 */

import { v } from 'convex/values';
import { mutation, query } from './_generated/server';

// ============================================
// 目的地存储
// ============================================

/**
 * 插入或更新目的地
 */
export const upsertDestination = mutation({
  args: {
    mddId: v.string(),
    sourceUrl: v.string(),
    name: v.string(),
    nameEn: v.optional(v.string()),
    country: v.optional(v.string()),
    province: v.optional(v.string()),
    description: v.optional(v.string()),
    coverImageUrl: v.optional(v.string()),
    imageUrls: v.array(v.string()),
    latitude: v.optional(v.number()),
    longitude: v.optional(v.number()),
    timezone: v.optional(v.string()),
    bestTravelTime: v.optional(v.string()),
    avgStayDays: v.optional(v.string()),
    climate: v.optional(v.string()),
    language: v.optional(v.string()),
    currency: v.optional(v.string()),
    visa: v.optional(v.string()),
    travelNotesCount: v.number(),
    poisCount: v.number(),
    questionsCount: v.number(),
    guidesCount: v.optional(v.number()),
  },
  handler: async (ctx, args) => {
    // 检查是否已存在
    const existing = await ctx.db
      .query('mafengwoDestinations')
      .withIndex('by_mdd_id', q => q.eq('mddId', args.mddId))
      .first();

    const now = Date.now();

    if (existing) {
      // 更新
      await ctx.db.patch(existing._id, {
        ...args,
        updatedAt: now,
      });
      return existing._id;
    }
    else {
      // 插入
      return await ctx.db.insert('mafengwoDestinations', {
        ...args,
        crawledAt: now,
      });
    }
  },
});

// ============================================
// POI 存储
// ============================================

/**
 * 插入或更新 POI
 */
export const upsertPoi = mutation({
  args: {
    poiId: v.string(),
    sourceUrl: v.string(),
    name: v.string(),
    nameEn: v.optional(v.string()),
    category: v.union(
      v.literal('attraction'),
      v.literal('restaurant'),
      v.literal('hotel'),
      v.literal('shopping'),
      v.literal('entertainment'),
      v.literal('transport'),
    ),
    destinationId: v.optional(v.string()),
    destinationName: v.optional(v.string()),
    address: v.optional(v.string()),
    latitude: v.optional(v.number()),
    longitude: v.optional(v.number()),
    rating: v.optional(v.number()),
    ratingCount: v.number(),
    priceLevel: v.optional(v.number()),
    priceRange: v.optional(v.string()),
    ticketPrice: v.optional(v.string()),
    openingHours: v.optional(v.string()),
    phone: v.optional(v.string()),
    website: v.optional(v.string()),
    description: v.optional(v.string()),
    tips: v.array(v.string()),
    highlights: v.array(v.string()),
    coverImageUrl: v.optional(v.string()),
    imageUrls: v.array(v.string()),
    reviewsCount: v.number(),
    savesCount: v.number(),
    tags: v.array(v.string()),
    cuisineType: v.optional(v.string()),
    signatureDishes: v.array(v.string()),
    starRating: v.optional(v.number()),
    amenities: v.array(v.string()),
    checkInTime: v.optional(v.string()),
    checkOutTime: v.optional(v.string()),
    qualityScore: v.number(),
  },
  handler: async (ctx, args) => {
    // 检查是否已存在
    const existing = await ctx.db
      .query('mafengwoPois')
      .withIndex('by_poi_id', q => q.eq('poiId', args.poiId))
      .first();

    const now = Date.now();

    if (existing) {
      // 更新
      await ctx.db.patch(existing._id, {
        ...args,
        updatedAt: now,
      });
      return existing._id;
    }
    else {
      // 插入
      return await ctx.db.insert('mafengwoPois', {
        ...args,
        crawledAt: now,
      });
    }
  },
});

/**
 * 批量插入 POI
 */
export const batchInsertPois = mutation({
  args: {
    pois: v.array(v.object({
      poiId: v.string(),
      sourceUrl: v.string(),
      name: v.string(),
      category: v.string(),
      destinationId: v.optional(v.string()),
      destinationName: v.optional(v.string()),
      rating: v.optional(v.number()),
      coverImageUrl: v.optional(v.string()),
      address: v.optional(v.string()),
    })),
  },
  handler: async (ctx, args) => {
    const now = Date.now();
    let inserted = 0;
    let updated = 0;

    for (const poi of args.pois) {
      const existing = await ctx.db
        .query('mafengwoPois')
        .withIndex('by_poi_id', q => q.eq('poiId', poi.poiId))
        .first();

      if (existing) {
        // 更新基本信息
        await ctx.db.patch(existing._id, {
          name: poi.name,
          rating: poi.rating,
          coverImageUrl: poi.coverImageUrl,
          address: poi.address,
          updatedAt: now,
        });
        updated++;
      }
      else {
        // 插入新记录
        await ctx.db.insert('mafengwoPois', {
          poiId: poi.poiId,
          sourceUrl: poi.sourceUrl,
          name: poi.name,
          category: poi.category as 'attraction' | 'restaurant' | 'hotel' | 'shopping' | 'entertainment' | 'transport',
          destinationId: poi.destinationId,
          destinationName: poi.destinationName,
          rating: poi.rating,
          ratingCount: 0,
          coverImageUrl: poi.coverImageUrl,
          imageUrls: [],
          address: poi.address,
          tips: [],
          highlights: [],
          reviewsCount: 0,
          savesCount: 0,
          tags: [],
          signatureDishes: [],
          amenities: [],
          qualityScore: 0.3, // 初始低质量分数
          crawledAt: now,
        });
        inserted++;
      }
    }

    return { inserted, updated, total: args.pois.length };
  },
});

// ============================================
// 攻略存储
// ============================================

/**
 * 插入或更新攻略
 */
export const upsertGuide = mutation({
  args: {
    guideId: v.string(),
    sourceUrl: v.string(),
    title: v.string(),
    destinationId: v.optional(v.string()),
    destinationName: v.optional(v.string()),
    authorName: v.optional(v.string()),
    authorId: v.optional(v.string()),
    summary: v.optional(v.string()),
    content: v.string(),
    contentHtml: v.optional(v.string()),
    sections: v.array(v.object({
      title: v.string(),
      content: v.string(),
      order: v.number(),
    })),
    coverImageUrl: v.optional(v.string()),
    imageUrls: v.array(v.string()),
    viewsCount: v.number(),
    likesCount: v.number(),
    savesCount: v.number(),
    commentsCount: v.number(),
    tags: v.array(v.string()),
    publishedAt: v.optional(v.number()),
    qualityScore: v.number(),
  },
  handler: async (ctx, args) => {
    // 检查是否已存在
    const existing = await ctx.db
      .query('mafengwoGuides')
      .withIndex('by_guide_id', q => q.eq('guideId', args.guideId))
      .first();

    const now = Date.now();

    if (existing) {
      // 更新
      await ctx.db.patch(existing._id, {
        ...args,
        updatedAt: now,
      });
      return existing._id;
    }
    else {
      // 插入
      return await ctx.db.insert('mafengwoGuides', {
        ...args,
        crawledAt: now,
      });
    }
  },
});

// ============================================
// 问答存储
// ============================================

/**
 * 插入或更新问答
 */
export const upsertQa = mutation({
  args: {
    questionId: v.string(),
    sourceUrl: v.string(),
    title: v.string(),
    content: v.string(),
    destinationId: v.optional(v.string()),
    destinationName: v.optional(v.string()),
    authorName: v.optional(v.string()),
    authorId: v.optional(v.string()),
    answersCount: v.number(),
    viewsCount: v.number(),
    tags: v.array(v.string()),
    createdAt: v.optional(v.number()),
    bestAnswer: v.optional(v.object({
      content: v.string(),
      authorName: v.optional(v.string()),
      authorId: v.optional(v.string()),
      likesCount: v.number(),
      createdAt: v.optional(v.number()),
    })),
  },
  handler: async (ctx, args) => {
    // 检查是否已存在
    const existing = await ctx.db
      .query('mafengwoQa')
      .withIndex('by_question_id', q => q.eq('questionId', args.questionId))
      .first();

    const now = Date.now();

    if (existing) {
      // 更新
      await ctx.db.patch(existing._id, {
        ...args,
      });
      return existing._id;
    }
    else {
      // 插入
      return await ctx.db.insert('mafengwoQa', {
        ...args,
        crawledAt: now,
      });
    }
  },
});

// ============================================
// 评论存储
// ============================================

/**
 * 插入或更新评论
 */
export const upsertReview = mutation({
  args: {
    reviewId: v.string(),
    poiExternalId: v.string(),
    poiName: v.optional(v.string()),
    authorName: v.optional(v.string()),
    authorId: v.optional(v.string()),
    authorAvatarUrl: v.optional(v.string()),
    rating: v.optional(v.number()),
    content: v.string(),
    imageUrls: v.array(v.string()),
    likesCount: v.number(),
    tags: v.array(v.string()),
    visitDate: v.optional(v.string()),
    createdAt: v.optional(v.number()),
  },
  handler: async (ctx, args) => {
    // 检查是否已存在
    const existing = await ctx.db
      .query('mafengwoReviews')
      .withIndex('by_review_id', q => q.eq('reviewId', args.reviewId))
      .first();

    const now = Date.now();

    if (existing) {
      // 更新
      await ctx.db.patch(existing._id, {
        ...args,
      });
      return existing._id;
    }
    else {
      // 插入
      return await ctx.db.insert('mafengwoReviews', {
        ...args,
        crawledAt: now,
      });
    }
  },
});

// ============================================
// 榜单存储
// ============================================

/**
 * 插入或更新榜单
 */
export const upsertRanking = mutation({
  args: {
    rankingId: v.string(),
    sourceUrl: v.string(),
    rankingType: v.union(
      v.literal('must_visit'),
      v.literal('food'),
      v.literal('hotel'),
      v.literal('shopping'),
      v.literal('hidden_gem'),
    ),
    title: v.string(),
    destinationId: v.optional(v.string()),
    destinationName: v.optional(v.string()),
    description: v.optional(v.string()),
    items: v.array(v.object({
      rank: v.number(),
      poiExternalId: v.string(),
      name: v.string(),
      category: v.optional(v.string()),
      rating: v.optional(v.number()),
      reviewsCount: v.number(),
      coverImageUrl: v.optional(v.string()),
      reason: v.optional(v.string()),
    })),
  },
  handler: async (ctx, args) => {
    // 检查是否已存在
    const existing = await ctx.db
      .query('mafengwoRankings')
      .withIndex('by_ranking_id', q => q.eq('rankingId', args.rankingId))
      .first();

    const now = Date.now();

    if (existing) {
      // 更新
      await ctx.db.patch(existing._id, {
        ...args,
        updatedAt: now,
      });
      return existing._id;
    }
    else {
      // 插入
      return await ctx.db.insert('mafengwoRankings', {
        ...args,
        crawledAt: now,
      });
    }
  },
});

// ============================================
// 爬取任务存储
// ============================================

/**
 * 创建爬取任务
 */
export const createCrawlTask = mutation({
  args: {
    taskType: v.union(
      v.literal('destination_list'),
      v.literal('destination_detail'),
      v.literal('poi_list'),
      v.literal('poi_detail'),
      v.literal('guide_list'),
      v.literal('guide_detail'),
      v.literal('travel_note_list'),
      v.literal('travel_note_detail'),
      v.literal('qa_list'),
      v.literal('qa_detail'),
      v.literal('ranking'),
      v.literal('review_list'),
    ),
    config: v.object({
      destinationId: v.optional(v.string()),
      destinationName: v.optional(v.string()),
      poiCategory: v.optional(v.string()),
      rankingType: v.optional(v.string()),
      targetUrl: v.optional(v.string()),
      page: v.optional(v.number()),
      pageSize: v.optional(v.number()),
      scrollCount: v.optional(v.number()),
      useAI: v.optional(v.boolean()),
    }),
    priority: v.number(),
  },
  handler: async (ctx, args) => {
    const now = Date.now();

    return await ctx.db.insert('mafengwoCrawlTasks', {
      taskType: args.taskType,
      config: args.config,
      priority: args.priority,
      status: 'pending',
      retryCount: 0,
      maxRetries: 3,
      createdAt: now,
    });
  },
});

/**
 * 更新任务状态
 */
export const updateCrawlTaskStatus = mutation({
  args: {
    taskId: v.id('mafengwoCrawlTasks'),
    status: v.union(
      v.literal('pending'),
      v.literal('running'),
      v.literal('completed'),
      v.literal('failed'),
      v.literal('cancelled'),
    ),
    itemsCollected: v.optional(v.number()),
    errorMessage: v.optional(v.string()),
  },
  handler: async (ctx, args) => {
    const now = Date.now();

    const updates: Record<string, unknown> = {
      status: args.status,
    };

    if (args.status === 'running') {
      updates.startedAt = now;
    }
    else if (args.status === 'completed' || args.status === 'failed') {
      updates.completedAt = now;
    }

    if (args.itemsCollected !== undefined) {
      updates.itemsCollected = args.itemsCollected;
    }

    if (args.errorMessage !== undefined) {
      updates.errorMessage = args.errorMessage;
    }

    await ctx.db.patch(args.taskId, updates);
  },
});

/**
 * 获取待处理任务
 */
export const getPendingTasks = query({
  args: {
    limit: v.optional(v.number()),
  },
  handler: async (ctx, args) => {
    const limit = args.limit || 10;

    return await ctx.db
      .query('mafengwoCrawlTasks')
      .withIndex('by_status_priority', q => q.eq('status', 'pending'))
      .order('desc')
      .take(limit);
  },
});

// ============================================
// 查询接口
// ============================================

/**
 * 获取目的地列表
 */
export const listDestinations = query({
  args: {
    limit: v.optional(v.number()),
    country: v.optional(v.string()),
  },
  handler: async (ctx, args) => {
    const limit = args.limit || 50;

    if (args.country) {
      return await ctx.db
        .query('mafengwoDestinations')
        .withIndex('by_country', q => q.eq('country', args.country))
        .take(limit);
    }

    return await ctx.db
      .query('mafengwoDestinations')
      .order('desc')
      .take(limit);
  },
});

/**
 * 获取目的地详情
 */
export const getDestination = query({
  args: {
    mddId: v.string(),
  },
  handler: async (ctx, args) => {
    return await ctx.db
      .query('mafengwoDestinations')
      .withIndex('by_mdd_id', q => q.eq('mddId', args.mddId))
      .first();
  },
});

/**
 * 获取 POI 列表
 */
export const listPois = query({
  args: {
    destinationId: v.optional(v.string()),
    category: v.optional(v.string()),
    limit: v.optional(v.number()),
  },
  handler: async (ctx, args) => {
    const limit = args.limit || 50;

    if (args.destinationId && args.category) {
      return await ctx.db
        .query('mafengwoPois')
        .withIndex('by_destination_category', q =>
          q.eq('destinationId', args.destinationId).eq('category', args.category as 'attraction'))
        .take(limit);
    }

    if (args.destinationId) {
      return await ctx.db
        .query('mafengwoPois')
        .withIndex('by_destination', q => q.eq('destinationId', args.destinationId))
        .take(limit);
    }

    if (args.category) {
      return await ctx.db
        .query('mafengwoPois')
        .withIndex('by_category', q => q.eq('category', args.category as 'attraction'))
        .take(limit);
    }

    return await ctx.db
      .query('mafengwoPois')
      .order('desc')
      .take(limit);
  },
});

/**
 * 获取攻略列表
 */
export const listGuides = query({
  args: {
    destinationId: v.optional(v.string()),
    limit: v.optional(v.number()),
  },
  handler: async (ctx, args) => {
    const limit = args.limit || 50;

    if (args.destinationId) {
      return await ctx.db
        .query('mafengwoGuides')
        .withIndex('by_destination', q => q.eq('destinationId', args.destinationId))
        .take(limit);
    }

    return await ctx.db
      .query('mafengwoGuides')
      .withIndex('by_views')
      .order('desc')
      .take(limit);
  },
});

/**
 * 获取榜单
 */
export const getRanking = query({
  args: {
    destinationId: v.string(),
    rankingType: v.string(),
  },
  handler: async (ctx, args) => {
    return await ctx.db
      .query('mafengwoRankings')
      .withIndex('by_destination_type', q =>
        q.eq('destinationId', args.destinationId).eq('rankingType', args.rankingType as 'must_visit'))
      .first();
  },
});

/**
 * 统计数据
 */
export const getStats = query({
  args: {},
  handler: async (ctx) => {
    const destinations = await ctx.db.query('mafengwoDestinations').collect();
    const pois = await ctx.db.query('mafengwoPois').collect();
    const guides = await ctx.db.query('mafengwoGuides').collect();
    const qa = await ctx.db.query('mafengwoQa').collect();
    const reviews = await ctx.db.query('mafengwoReviews').collect();
    const rankings = await ctx.db.query('mafengwoRankings').collect();
    const tasks = await ctx.db.query('mafengwoCrawlTasks').collect();

    const pendingTasks = tasks.filter(t => t.status === 'pending').length;
    const runningTasks = tasks.filter(t => t.status === 'running').length;
    const completedTasks = tasks.filter(t => t.status === 'completed').length;
    const failedTasks = tasks.filter(t => t.status === 'failed').length;

    return {
      destinations: destinations.length,
      pois: pois.length,
      guides: guides.length,
      qa: qa.length,
      reviews: reviews.length,
      rankings: rankings.length,
      tasks: {
        total: tasks.length,
        pending: pendingTasks,
        running: runningTasks,
        completed: completedTasks,
        failed: failedTasks,
      },
    };
  },
});

File: apps/server/internal/model/mafengwo.go
package model

import "time"

// JSONStringSlice is a helper type for database JSON columns that store
// a JSON array of strings (e.g. TEXT/JSON columns).
type JSONStringSlice []string

// ---------------------------------------------------------------------------
// MafengwoDestination — a crawled destination page from mafengwo.cn
// ---------------------------------------------------------------------------

type MafengwoDestination struct {
	ID               int64      `json:"id" db:"id"`
	MddID            string     `json:"mddId" db:"mdd_id"`
	SourceURL        string     `json:"sourceUrl" db:"source_url"`
	Name             string     `json:"name" db:"name"`
	NameEn           *string    `json:"nameEn,omitempty" db:"name_en"`
	Country          *string    `json:"country,omitempty" db:"country"`
	Province         *string    `json:"province,omitempty" db:"province"`
	Description      *string    `json:"description,omitempty" db:"description"`
	CoverImageURL    *string    `json:"coverImageUrl,omitempty" db:"cover_image_url"`
	ImageURLs        []string   `json:"imageUrls" db:"image_urls"`
	Latitude         *float64   `json:"latitude,omitempty" db:"latitude"`
	Longitude        *float64   `json:"longitude,omitempty" db:"longitude"`
	Timezone         *string    `json:"timezone,omitempty" db:"timezone"`
	BestTravelTime   *string    `json:"bestTravelTime,omitempty" db:"best_travel_time"`
	AvgStayDays      *string    `json:"avgStayDays,omitempty" db:"avg_stay_days"`
	Climate          *string    `json:"climate,omitempty" db:"climate"`
	Language         *string    `json:"language,omitempty" db:"language"`
	Currency         *string    `json:"currency,omitempty" db:"currency"`
	Visa             *string    `json:"visa,omitempty" db:"visa"`
	TravelNotesCount int        `json:"travelNotesCount" db:"travel_notes_count"`
	PoisCount        int        `json:"poisCount" db:"pois_count"`
	QuestionsCount   int        `json:"questionsCount" db:"questions_count"`
	GuidesCount      *int       `json:"guidesCount,omitempty" db:"guides_count"`
	CrawledAt        time.Time  `json:"crawledAt" db:"crawled_at"`
	UpdatedAt        *time.Time `json:"updatedAt,omitempty" db:"updated_at"`
}

// ---------------------------------------------------------------------------
// MafengwoPOI — a point-of-interest (attraction / restaurant / hotel / shopping)
// ---------------------------------------------------------------------------

type MafengwoPOI struct {
	ID              int64      `json:"id" db:"id"`
	PoiID           string     `json:"poiId" db:"poi_id"`
	SourceURL       string     `json:"sourceUrl" db:"source_url"`
	Name            string     `json:"name" db:"name"`
	NameEn          *string    `json:"nameEn,omitempty" db:"name_en"`
	Category        string     `json:"category" db:"category"` // attraction / restaurant / hotel / shopping
	DestinationID   *string    `json:"destinationId,omitempty" db:"destination_id"`
	DestinationName *string    `json:"destinationName,omitempty" db:"destination_name"`
	Address         *string    `json:"address,omitempty" db:"address"`
	Latitude        *float64   `json:"latitude,omitempty" db:"latitude"`
	Longitude       *float64   `json:"longitude,omitempty" db:"longitude"`
	Rating          *float64   `json:"rating,omitempty" db:"rating"`
	RatingCount     int        `json:"ratingCount" db:"rating_count"`
	PriceLevel      *int       `json:"priceLevel,omitempty" db:"price_level"`
	PriceRange      *string    `json:"priceRange,omitempty" db:"price_range"`
	TicketPrice     *string    `json:"ticketPrice,omitempty" db:"ticket_price"`
	OpeningHours    *string    `json:"openingHours,omitempty" db:"opening_hours"`
	Phone           *string    `json:"phone,omitempty" db:"phone"`
	Website         *string    `json:"website,omitempty" db:"website"`
	Description     *string    `json:"description,omitempty" db:"description"`
	Tips            []string   `json:"tips" db:"tips"`
	Highlights      []string   `json:"highlights" db:"highlights"`
	CoverImageURL   *string    `json:"coverImageUrl,omitempty" db:"cover_image_url"`
	ImageURLs       []string   `json:"imageUrls" db:"image_urls"`
	ReviewsCount    int        `json:"reviewsCount" db:"reviews_count"`
	SavesCount      int        `json:"savesCount" db:"saves_count"`
	Tags            []string   `json:"tags" db:"tags"`
	CuisineType     *string    `json:"cuisineType,omitempty" db:"cuisine_type"`
	SignatureDishes []string   `json:"signatureDishes" db:"signature_dishes"`
	StarRating      *int       `json:"starRating,omitempty" db:"star_rating"`
	Amenities       []string   `json:"amenities" db:"amenities"`
	CheckInTime     *string    `json:"checkInTime,omitempty" db:"check_in_time"`
	CheckOutTime    *string    `json:"checkOutTime,omitempty" db:"check_out_time"`
	QualityScore    int        `json:"qualityScore" db:"quality_score"`
	CrawledAt       time.Time  `json:"crawledAt" db:"crawled_at"`
	UpdatedAt       *time.Time `json:"updatedAt,omitempty" db:"updated_at"`
}

// ---------------------------------------------------------------------------
// MafengwoGuide — a travel guide / "gonglue" article
// ---------------------------------------------------------------------------

type MafengwoGuide struct {
	ID              int64      `json:"id" db:"id"`
	GuideID         string     `json:"guideId" db:"guide_id"`
	SourceURL       string     `json:"sourceUrl" db:"source_url"`
	Title           string     `json:"title" db:"title"`
	DestinationID   *string    `json:"destinationId,omitempty" db:"destination_id"`
	DestinationName *string    `json:"destinationName,omitempty" db:"destination_name"`
	AuthorName      *string    `json:"authorName,omitempty" db:"author_name"`
	AuthorURL       *string    `json:"authorUrl,omitempty" db:"author_url"`
	Content         *string    `json:"content,omitempty" db:"content"`
	Summary         *string    `json:"summary,omitempty" db:"summary"`
	CoverImageURL   *string    `json:"coverImageUrl,omitempty" db:"cover_image_url"`
	ImageURLs       []string   `json:"imageUrls" db:"image_urls"`
	Tags            []string   `json:"tags" db:"tags"`
	ViewCount       int        `json:"viewCount" db:"view_count"`
	LikeCount       int        `json:"likeCount" db:"like_count"`
	CommentCount    int        `json:"commentCount" db:"comment_count"`
	ShareCount      int        `json:"shareCount" db:"share_count"`
	PublishedAt     *time.Time `json:"publishedAt,omitempty" db:"published_at"`
	CrawledAt       time.Time  `json:"crawledAt" db:"crawled_at"`
	UpdatedAt       *time.Time `json:"updatedAt,omitempty" db:"updated_at"`
}

// ---------------------------------------------------------------------------
// MafengwoQA — a question & answer from the Q&A section
// ---------------------------------------------------------------------------

type BestAnswer struct {
	AuthorName string `json:"authorName" db:"best_answer_author"`
	Content    string `json:"content" db:"best_answer_content"`
	LikeCount  int    `json:"likeCount" db:"best_answer_likes"`
}

type MafengwoQA struct {
	ID              int64       `json:"id" db:"id"`
	QuestionID      string      `json:"questionId" db:"question_id"`
	SourceURL       string      `json:"sourceUrl" db:"source_url"`
	Title           string      `json:"title" db:"title"`
	Content         *string     `json:"content,omitempty" db:"content"`
	DestinationID   *string     `json:"destinationId,omitempty" db:"destination_id"`
	DestinationName *string     `json:"destinationName,omitempty" db:"destination_name"`
	AuthorName      *string     `json:"authorName,omitempty" db:"author_name"`
	Tags            []string    `json:"tags" db:"tags"`
	AnswerCount     int         `json:"answerCount" db:"answer_count"`
	ViewCount       int         `json:"viewCount" db:"view_count"`
	BestAnswer      *BestAnswer `json:"bestAnswer,omitempty" db:"best_answer"`
	CrawledAt       time.Time   `json:"crawledAt" db:"crawled_at"`
	UpdatedAt       *time.Time  `json:"updatedAt,omitempty" db:"updated_at"`
}

// ---------------------------------------------------------------------------
// MafengwoRanking — a ranking / leaderboard page
// ---------------------------------------------------------------------------

type RankingItem struct {
	Rank          int      `json:"rank"`
	Name          string   `json:"name"`
	PoiID         *string  `json:"poiId,omitempty"`
	Score         *float64 `json:"score,omitempty"`
	CoverImageURL *string  `json:"coverImageUrl,omitempty"`
	Description   *string  `json:"description,omitempty"`
}

type MafengwoRanking struct {
	ID              int64         `json:"id" db:"id"`
	RankingID       string        `json:"rankingId" db:"ranking_id"`
	SourceURL       string        `json:"sourceUrl" db:"source_url"`
	Title           string        `json:"title" db:"title"`
	Category        string        `json:"category" db:"category"`
	DestinationID   *string       `json:"destinationId,omitempty" db:"destination_id"`
	DestinationName *string       `json:"destinationName,omitempty" db:"destination_name"`
	Items           []RankingItem `json:"items" db:"items"`
	TotalItems      int           `json:"totalItems" db:"total_items"`
	CrawledAt       time.Time     `json:"crawledAt" db:"crawled_at"`
	UpdatedAt       *time.Time    `json:"updatedAt,omitempty" db:"updated_at"`
}

// ---------------------------------------------------------------------------
// MafengwoCrawlTask — tracks crawl job state
// ---------------------------------------------------------------------------

type MafengwoCrawlTask struct {
	ID           int64      `json:"id" db:"id"`
	TaskType     string     `json:"taskType" db:"task_type"` // destination / poi / guide / qa / ranking
	TargetURL    string     `json:"targetUrl" db:"target_url"`
	Status       string     `json:"status" db:"status"` // pending / running / completed / failed
	Priority     int        `json:"priority" db:"priority"`
	RetryCount   int        `json:"retryCount" db:"retry_count"`
	MaxRetries   int        `json:"maxRetries" db:"max_retries"`
	ErrorMessage *string    `json:"errorMessage,omitempty" db:"error_message"`
	ResultID     *int64     `json:"resultId,omitempty" db:"result_id"`
	ScheduledAt  *time.Time `json:"scheduledAt,omitempty" db:"scheduled_at"`
	StartedAt    *time.Time `json:"startedAt,omitempty" db:"started_at"`
	CompletedAt  *time.Time `json:"completedAt,omitempty" db:"completed_at"`
	CreatedAt    time.Time  `json:"createdAt" db:"created_at"`
	UpdatedAt    *time.Time `json:"updatedAt,omitempty" db:"updated_at"`
}

// ---------------------------------------------------------------------------
// TravelGuide — a synthesized / AI-generated travel guide
// ---------------------------------------------------------------------------

type GuideDestination struct {
	Name      string   `json:"name"`
	MddID     *string  `json:"mddId,omitempty"`
	Country   *string  `json:"country,omitempty"`
	Province  *string  `json:"province,omitempty"`
	Latitude  *float64 `json:"latitude,omitempty"`
	Longitude *float64 `json:"longitude,omitempty"`
}

type DayItinerary struct {
	Day           int      `json:"day"`
	Title         string   `json:"title"`
	Description   *string  `json:"description,omitempty"`
	POIs          []string `json:"pois"`
	Meals         []string `json:"meals,omitempty"`
	Accommodation *string  `json:"accommodation,omitempty"`
	Transport     *string  `json:"transport,omitempty"`
	Tips          []string `json:"tips,omitempty"`
}

type TravelGuide struct {
	ID            int64              `json:"id" db:"id"`
	ExternalID    string             `json:"externalId" db:"external_id"`
	Platform      string             `json:"platform" db:"platform"` // mafengwo / ai-generated
	Title         string             `json:"title" db:"title"`
	Content       *string            `json:"content,omitempty" db:"content"`
	AuthorName    *string            `json:"authorName,omitempty" db:"author_name"`
	SourceURL     string             `json:"sourceUrl" db:"source_url"`
	CoverImageURL *string            `json:"coverImageUrl,omitempty" db:"cover_image_url"`
	ImageURLs     []string           `json:"imageUrls" db:"image_urls"`
	Destinations  []GuideDestination `json:"destinations" db:"destinations"`
	Tags          []string           `json:"tags" db:"tags"`
	TravelDays    *int               `json:"travelDays,omitempty" db:"travel_days"`
	TravelMonth   *string            `json:"travelMonth,omitempty" db:"travel_month"`
	Budget        *string            `json:"budget,omitempty" db:"budget"`
	TravelStyle   *string            `json:"travelStyle,omitempty" db:"travel_style"`
	Itinerary     []DayItinerary     `json:"itinerary" db:"itinerary"`
	ViewCount     int                `json:"viewCount" db:"view_count"`
	LikeCount     int                `json:"likeCount" db:"like_count"`
	QualityScore  int                `json:"qualityScore" db:"quality_score"`
	PublishedAt   *time.Time         `json:"publishedAt,omitempty" db:"published_at"`
	LastUpdatedAt *time.Time         `json:"lastUpdatedAt,omitempty" db:"last_updated_at"`
	CreatedAt     time.Time          `json:"createdAt" db:"created_at"`
	UpdatedAt     *time.Time         `json:"updatedAt,omitempty" db:"updated_at"`
}
